File: Leetcode_152.py

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_prod = nums[0]
        curr_max = nums[0]
        curr_min = nums[0]
        
        for i in range(1, len(nums)):
            num = nums[i]
            
            # If current number is negative, swap max and min
            if num < 0:
                curr_max, curr_min = curr_min, curr_max
            
            curr_max = max(num, num * curr_max)
            curr_min = min(num, num * curr_min)
            
            max_prod = max(max_prod, curr_max)
        
        return max_prod

File: Leetcode_152.java
class Solution {
    public int maxProduct(int[] nums) {
        int maxProd = nums[0];
        int currMax = nums[0];
        int currMin = nums[0];
        
        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];
            
            // If current number is negative, swap currMax and currMin
            if (num < 0) {
                int temp = currMax;
                currMax = currMin;
                currMin = temp;
            }
            
            currMax = Math.max(num, num * currMax);
            currMin = Math.min(num, num * currMin);
            
            maxProd = Math.max(maxProd, currMax);
        }
        
        return maxProd;
    }
}
